runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -0,0 +1,228 @@
+      do {
+        previous = logicalValue.get();
+      } while (!logicalValue.compareAndSet(previous, previous.add(update)));
+    }
+
+    @Override
+    public DistributionResult extractLogical() {
+      return logicalValue.get().extractResult();
+    }
+  }
+
+  /** The current values of counters in memory. */
+  private MetricsMap<MetricKey, DirectMetric<Long, Long>> counters =
+      new MetricsMap<MetricKey, DirectMetric<Long, Long>>() {
+        @Override
+        protected DirectMetric<Long, Long> createInstance() {
+          return new DirectCounter();
+        }
+      };
+  private MetricsMap<MetricKey, DirectMetric<DistributionData, DistributionResult>> distributions =
+      new MetricsMap<MetricKey, DirectMetric<DistributionData, DistributionResult>>() {
+        @Override
+        protected DirectMetric<DistributionData, DistributionResult> createInstance() {
+          return new DirectDistribution();
+        }
+      };
+
+  @AutoValue
+  abstract static class DirectMetricQueryResults implements MetricQueryResults {
+    public static MetricQueryResults create(
+        Iterable<MetricResult<Long>> counters,
+        Iterable<MetricResult<DistributionResult>> distributions) {
+      return new AutoValue_DirectMetrics_DirectMetricQueryResults(counters, distributions);
+    }
+  }
+
+  @AutoValue
+  abstract static class DirectMetricResult<T> implements MetricResult<T> {
+    public static <T> MetricResult<T> create(MetricName name, String scope,
+        T committed, T attempted) {
+      return new AutoValue_DirectMetrics_DirectMetricResult<T>(
+          name, scope, committed, attempted);
+    }
+  }
+
+  @Override
+  public MetricQueryResults queryMetrics(MetricFilter filter) {
+    ImmutableList.Builder<MetricResult<Long>> counterResults = ImmutableList.builder();
+    for (Entry<MetricKey, DirectMetric<Long, Long>> counter : counters.entries()) {
+      maybeExtractResult(filter, counterResults, counter);
+    }
+    ImmutableList.Builder<MetricResult<DistributionResult>> distributionResults =
+        ImmutableList.builder();
+    for (Entry<MetricKey, DirectMetric<DistributionData, DistributionResult>> distribution
+        : distributions.entries()) {
+      maybeExtractResult(filter, distributionResults, distribution);
+    }
+
+    return DirectMetricQueryResults.create(counterResults.build(), distributionResults.build());
+  }
+
+  private <ResultT> void maybeExtractResult(
+      MetricFilter filter,
+      ImmutableList.Builder<MetricResult<ResultT>> resultsBuilder,
+      Map.Entry<MetricKey, ? extends DirectMetric<?, ResultT>> entry) {
+    if (matches(filter, entry.getKey())) {
+      resultsBuilder.add(DirectMetricResult.create(
+          entry.getKey().metricName(),
+          entry.getKey().stepName(),
+          entry.getValue().extractLogical(),
+          entry.getValue().extractPhysical()));
+    }
+  }
+
+  private boolean matches(MetricFilter filter, MetricKey key) {
```

Comment:
See responses elsewhere about trying to keep the "protocol"-like classes simple, so they can be swapped out using an actual protocol.